src/abstractdataframe/join.jl

```diff
@@ -91,6 +91,305 @@ _rename_cols(old_names::AbstractVector{Symbol},
+            push!(left_ixs, idx_l)
+            push!(right_ixs, idx_r)
+        end
+    end
+    return left_ixs, right_ixs
+end
+
+# we fall back to general case if we have duplicates
+# normally it should happen fast as we reuse work already done
+function _innerjoin_dup(left::AbstractArray, right::AbstractArray{T}, dict::Dict{T, Int}, idx_r_start::Int) where {T}
+    ngroups = idx_r_start - 1
+    right_len = length(right)
+    groups = Vector{Int}(undef, right_len)
+    groups[1:ngroups] = 1:ngroups
+
+    for idx_r in idx_r_start:right_len
+        @inbounds val_r = right[idx_r]
+        # we use dict_index to make sure the following two operations are fast:
+        # - if index is found - process the row with existing group number
+        # - if index is not found - add a new group
+        dict_index = Base.ht_keyindex2!(dict, val_r)
+        if dict_index > 0
+            @inbounds groups[idx_r] = dict.vals[dict_index]
+        else
+            ngroups += 1
+            @inbounds groups[idx_r] = ngroups
+            Base._setindex!(dict, ngroups, val_r, -dict_index)
+        end
+    end
+
+    @assert ngroups > 0 # we should not get here with 0-length right
+    return _innerjoin_postprocess(left, dict, groups, ngroups, right_len)
+end
+
+function compute_join_indices!(groups::Vector{Int}, ngroups::Int,
+                               starts::Vector, rperm::Vector)
+    @inbounds for gix in groups
+        starts[gix] += 1
+    end
+
+    cumsum!(starts, starts)
+
+    @inbounds for (i, gix) in enumerate(groups)
+        rperm[starts[gix]] = i
+        starts[gix] -= 1
+    end
+    push!(starts, length(groups))
+    return nothing
+end
+
+function _innerjoin_postprocess(left::AbstractArray, dict::Dict{T, Int},
+                                groups::Vector{Int}, ngroups::Int, right_len::Int) where {T}
+    starts = zeros(Int, ngroups)
+    rperm = Vector{Int}(undef, right_len)
+
+    left_ixs = Int[]
+    right_ixs = Int[]
+
+    # lower bound assuming we get matches
+    sizehint!(left_ixs, right_len)
+    sizehint!(right_ixs, right_len)
+
+    compute_join_indices!(groups, ngroups, starts, rperm)
+
+    n = 0
+    @inbounds for (idx_l, val_l) in enumerate(left)
+        # we use dict_index to make sure the following two operations are fast:
+        # - if index is found - get it and process it
+        # - if index is not found - do nothing
+        dict_index = Base.ht_keyindex(dict, val_l)
+        if dict_index > 0 # -1 if key not found
+            group_id = dict.vals[dict_index]
+            @inbounds ref_stop = starts[group_id + 1]
+            @inbounds l = ref_stop - starts[group_id]
+            newn = n + l
+            resize!(left_ixs, newn)
+            @simd for i in n+1:n+l
+                @inbounds left_ixs[i] = idx_l
+            end
+            resize!(right_ixs, newn)
+            @simd for i in 1:l
+                @inbounds right_ixs[n + i] = rperm[ref_stop - i + 1]
+            end
```

Comment:
```suggestion
            ref_stop = starts[group_id + 1]
            l = ref_stop - starts[group_id]
            newn = n + l
            resize!(left_ixs, newn)
            @simd for i in n+1:n+l
                left_ixs[i] = idx_l
            end
            resize!(right_ixs, newn)
            @simd for i in 1:l
                right_ixs[n + i] = rperm[ref_stop - i + 1]
            end
```
Also are you sure `@simd` makes a difference here? AFAIK it's only useful for floating point operations.

Reply:
you are probably right. I will remove them